File: inhouse/16mer.jl
# 786

function revcomp(c)
	if c == 'A' 
		'T'
	elseif c == 'C' 
		'G'
	elseif c == 'G' 
		'C'
	elseif c == 'T' 
		'A'
	else 
		c 
	end
end

function process(s)
	length(s)
end

function sym(s, k)
	for j in 1:div(k, 2)
		if s[j] != revcomp(s[k - j + 1])
			return 0
		end
	end
	1
end

function main(arg)
	f = open(arg)
	t1, t2, t3 = 0, 0, 0
	for l in eachline(f)
		k, stp = 16, 1
	  i = 1
	  while i + k - 1 <= length(l)
	    sk = l[i:i+k-1]
	    t1 += process(sk)
	   	t2 += 1
	   	t3 += sym(sk, k)
	    i += stp
	  end
	end
	println("$t1 $t2 $t3")
end

main(ARGS[1])

File: inhouse/16mer.py
# 786
# benchmarks:
# - reverse complementing 10g FASTA
# - printing all 16-mers from 10g FASTA
# - count number of CpG islands (e.g. all CG-only subseqs)

import sys

def revcomp(c):
   return ('A' if c == 'T' else \
          ('C' if c == 'G' else \
          ('G' if c == 'C' else \
          ('T' if c == 'A' else c))))

def process(k):
   return len(k)

def ksplit(s, k, step):
   i = 0
   while i + k <= len(s):
      yield s[i:i + k]
      i += step

def sym(s, k):
   return 1 if all(s[i] == revcomp(s[-i-1]) for i in range(k//2)) else 0

f = open(sys.argv[1], 'r')
total, t2, t3 = 0, 0, 0
k = 16
for l in f:
   for s in ksplit(l.strip(), k, 1):
      total += process(s)
      t2 += 1
      t3 += sym(s, k)
print total, t2, t3
